src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp

```diff
@@ -388,7 +430,82 @@ DarknetYOLOModelTrainer::DarknetYOLOModelTrainer(
           /* h_out   */ config_.output_height,
           /* w_out   */ config_.output_width,
           /* config  */ checkpoint.internal_config(),
-          /* weights */ checkpoint.internal_weights())) {}
+          /* weights */ checkpoint.internal_weights())),
+      training_augmenter_(
+          std::make_shared<DataAugmenter>(context->create_image_augmenter(
+              DarknetYOLOTrainingAugmentationOptions(
+                  checkpoint.config().batch_size,
+                  checkpoint.config().output_height,
+                  checkpoint.config().output_width)))),
+      inference_augmenter_(
+          std::make_shared<DataAugmenter>(context->create_image_augmenter(
+              DarknetYOLOInferenceAugmentationOptions(
+                  checkpoint.config().batch_size,
+                  checkpoint.config().output_height,
+                  checkpoint.config().output_width)))) {}
+
+std::shared_ptr<Publisher<TrainingOutputBatch>>
+DarknetYOLOModelTrainer::AsTrainingBatchPublisher(
+    std::unique_ptr<data_iterator> training_data, size_t batch_size,
+    int offset) {
+  // Wrap the data_iterator to incorporate into a Combine pipeline.
+  auto iterator = std::make_shared<DataIterator>(std::move(training_data),
+                                                 batch_size, offset);
+
+  // Define a lambda that applies EncodeDarknetYOLO to the raw annotations.
+  Config config = config_;
+  auto encoder = [config](InputBatch input_batch) {
+    return EncodeDarknetYOLO(
+        std::move(input_batch), config.output_height, config.output_width,
+        static_cast<int>(GetAnchorBoxes().size()), config.num_classes);
+  };
+
+  // Wrap the model_backend.
+  auto trainer = std::make_shared<DarknetYOLOBackendTrainingWrapper>(
+      backend_, BASE_LEARNING_RATE, config_.max_iterations);
+
+  // Construct the training pipeline.
+  return iterator->AsPublisher()
+      ->Map(training_augmenter_)
+      ->Map(encoder)
+      ->Map(trainer);
+}
+
+std::shared_ptr<Publisher<EncodedBatch>>
+DarknetYOLOModelTrainer::AsInferenceBatchPublisher(
+    std::unique_ptr<data_iterator> test_data, size_t batch_size,
+    float confidence_threshold, float iou_threshold) {
+  // Wrap the data_iterator to incorporate into a Combine pipeline.
+  auto iterator = std::make_shared<DataIterator>(std::move(test_data),
+                                                 batch_size, /* offset */ 0);
+
+  // No labels to encode. Just pass the annotations through for potential
+  // evaluation.
+  auto trivial_encoder = [](InputBatch input_batch) {
+    EncodedInputBatch result;
```

Comment:
Why are we doing this?

Reply:
The augmenter currently outputs an `InputBatch` but the model backend wants an `EncodedInputBatch`, which is basically `InputBatch` plus some stuff for loss labels that don't apply to the inference case. So we just need to copy all the contents of the `InputBatch` into the `EncodedInputBatch`, since there are no loss labels to encode.